client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -55,16 +55,16 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(HonoClientImpl.class);
 
-    private final Map<String, MessageSender> activeSenders = new ConcurrentHashMap<>();
-    private final Map<String, RequestResponseClient> activeRequestResponseClients = new ConcurrentHashMap<>();
-    private final Map<String, Boolean> creationLocks = new ConcurrentHashMap<>();
+    private final Map<String, MessageSender> activeSenders = new HashMap<>();
+    private final Map<String, RequestResponseClient> activeRequestResponseClients = new HashMap<>();
+    private final Map<String, Boolean> creationLocks = new HashMap<>();
     private final List<Handler<Void>> creationRequests = new ArrayList<>();
     private final AtomicBoolean connecting = new AtomicBoolean(false);
+    private final AtomicBoolean shuttingDown = new AtomicBoolean(false);
     private final ConnectionFactory connectionFactory;
     private final ClientConfigProperties clientConfigProperties;
     private final Vertx vertx;
-
-    private volatile boolean shutdown = false;
+    private final Object connectionLock = new Object();
```

Comment:
Just some thoughts from my side:
do we really need this Object?
Why not use `this` as a lock instead?
Since the object is always created iff HonoClient is created, `this` should be just as good, if I judge things correctly.

WDYT?

Reply:
You are probably right since we always only synchronize on the same semaphore.

Reply:
What I just read additionally:
sometimes your way of using a "private" lock is considered better, since no one outside the class can synchronize on this lock (since there is no access to it).
This is not the case when using "this" instead.

So, sometimes there are some subtle advantages in creating an own lock.
But this might be not so important here - WDYT?

Reply:
If you sync on `this` you effectively make this behavior part of your API. As @sysexcontrol mentioned, others can interfere with your lock here. IMHO keeping the explicit (but strictly internal) lock is the better way.

Reply:
:+1: